services/sleeper_task.go

```diff
@@ -19,55 +13,57 @@ type Worker interface {
 }
 
 type sleeperTask struct {
-	worker  Worker
-	store   *store.Store
-	cond    *sync.Cond
-	started bool
+	worker Worker
+	waker  chan struct{}
+	closer chan struct{}
 }
 
-// NewSleeperTask takes a worker and retruns a SleeperTask
+// NewSleeperTask takes a worker and returns a SleeperTask.
+//
+// SleeperTask is guaranteed to call Work on the worker at least once for every
+// WakeUp call.
+// If the Worker is busy when WakeUp is called, the Worker will be called again
+// immediately after it is finished. For this reason you should take care to
+// make sure that Worker is idempotent.
+// WakeUp does not block.
+//
 func NewSleeperTask(worker Worker) SleeperTask {
-	var m sync.Mutex
 	return &sleeperTask{
 		worker: worker,
-		cond:   sync.NewCond(&m),
+		waker:  make(chan struct{}, 1),
+		closer: make(chan struct{}),
 	}
 }
 
 // Start begins the SleeperTask
 func (s *sleeperTask) Start() error {
-	s.cond.L.Lock()
-	s.started = true
-	s.cond.L.Unlock()
 	go s.workerLoop()
 	return nil
 }
 
 // Stop stops the SleeperTask
 func (s *sleeperTask) Stop() error {
-	s.cond.L.Lock()
-	s.started = false
-	s.cond.Signal()
-	s.cond.L.Unlock()
+	s.closer <- struct{}{}
```

Comment:
If we `Start` and then `Stop` extremely quickly before the `go s.workerLoop()` can boot up, it'll miss the message and leak a goroutine. We could mitigate this issue by using a `sync.WaitGroup` in `Start` to block until the goroutine is ready.

Also, we don't seem resilient to a `Start`, `Stop`, `Start` scenario since we reuse closed channels. We could either not care and move on with life or initialize the channels in `Start`.

Reply:
I don't think (1) is true, it'll block the `Stop` routine until the `workerLoop` takes the `closer` message and then cleanup.

I'll add a test for (2).